tests/unittests/test_datasource/test_azure_helper.py

```diff
@@ -365,6 +459,113 @@ def test_parse_certificates(self, mock_decrypt_certs):
             self.assertIn(fp, keys_by_fp)
 
 
+class TestGoalStateHealthReporter(CiTestCase):
+
+    default_parameters = {
+        'incarnation': 1,
+        'container_id': 'MyContainerId',
+        'instance_id': 'MyInstanceId'
+    }
+
+    test_endpoint = 'TestEndpoint'
+    test_url = 'http://{0}/machine?comp=health'.format(test_endpoint)
+    test_default_headers = {'Content-Type': 'text/xml; charset=utf-8'}
+
+    provisioning_success_status = 'Ready'
+
+    def setUp(self):
+        super(TestGoalStateHealthReporter, self).setUp()
+        patches = ExitStack()
+        self.addCleanup(patches.close)
+
+        patches.enter_context(
+            mock.patch.object(azure_helper.time, 'sleep', mock.MagicMock()))
+        self.read_file_or_url = patches.enter_context(
+            mock.patch.object(azure_helper.url_helper, 'read_file_or_url'))
+
+        self.post = patches.enter_context(
+            mock.patch.object(azure_helper.AzureEndpointHttpClient,
+                              'post'))
+
+        self.GoalState = patches.enter_context(
+            mock.patch.object(azure_helper, 'GoalState'))
+        self.GoalState.return_value.container_id = \
+            self.default_parameters['container_id']
+        self.GoalState.return_value.instance_id = \
+            self.default_parameters['instance_id']
+        self.GoalState.return_value.incarnation = \
+            self.default_parameters['incarnation']
+
+    def _get_formatted_health_report_xml_string(self, **kwargs):
+        return HEALTH_REPORT_XML_TEMPLATE.format(**kwargs)
+
+    def _get_report_ready_health_document(self):
+        return self._get_formatted_health_report_xml_string(
+            incarnation=self.default_parameters['incarnation'],
+            container_id=self.default_parameters['container_id'],
+            instance_id=self.default_parameters['instance_id'],
+            health_status=self.provisioning_success_status,
+            health_detail_subsection='')
+
+    def test_send_ready_signal_sends_post_request(self):
```

Comment:
I feel like there's some overlap in these tests which could be cleaned up to make them a little clearer. This test implies it tests that "`send_ready_signal` sends a POST request", but it's also testing that `build_report` is called and that `build_report` behaves as expected; if not, the final assertion would fail here because `data=health_document` would not match. This encompasses the following two tests entirely, rendering them unnecessary.

`test_send_ready_signal_health_document` is already testing `build_report`, so we don't also need to test it here. I propose modifying this test to include the mocking of `build_report` from `test_send_ready_signal_calls_build_report`, and instead of `mock.call(..., data=health_document, ...)` instead test for `mock.call(..., data=m_build_report.return_value, ...)`. (I suggest keeping `test_send_ready_signal_calls_build_report` separate; validating the parameters can be performed separately, and the less each test is asserting, the easier it is to grok them.)